### PR TITLE
Issue 45298: Optimize notification count retrieval

### PR DESCRIPTION
**Pull request by labkey-jeckels**
#### Rationale
When the notification menu is turned on, it adds a DB query to every page load. While 10-20ms extra isn't terrible, it's a lot of time to add to every single page.

#### Changes
* Cache unread notification count on a per-user basis
* Clear cache when notifications come and go

### CODE REVIEW

File: core/src/org/labkey/core/notification/NotificationServiceImpl.java [hunk 1]
```diff
@@ -82,6 +86,8 @@ private NotificationServiceImpl()
         ContainerManager.addContainerListener(this);
     }
 
+    /** Cache the number of unread notifications per user (key is a user ID) to avoid a DB query on every page load */
+    private final Cache<Integer, Long> _unreadCountCache = CacheManager.getCache(CacheManager.UNLIMITED, TimeUnit.HOURS.toMillis(1), "Unread notification counts");
```
Comment (labkey-matthewb) [inline core/src/org/labkey/core/notification/NotificationServiceImpl.java]: Consider using BlockingCache

File: core/src/org/labkey/core/notification/NotificationServiceImpl.java [hunk 2]
```diff
@@ -175,21 +179,33 @@ public Notification addNotification(Container container, User user, @NotNull Not
...
         return getNotificationsByUserOrType(container, null, notifyUserId, unreadOnly);
     }
 
     @Override
-    public long getNotificationCountByUser(Container container, int notifyUserId, boolean unreadOnly)
+    public long getUnreadNotificationCountByUser(Container container, int notifyUserId)
     {
-        return createSelectorByUserOrType(container, null, notifyUserId, unreadOnly).getRowCount();
+        return _unreadCountCache.get(notifyUserId, null,
+                (k, a) -> createSelectorByUserOrType(container, null, notifyUserId, true).getRowCount());
```
Comment (labkey-matthewb) [inline core/src/org/labkey/core/notification/NotificationServiceImpl.java]: Doesn't address the fact that core.notifications doesn't have a useful index that matches the query generated by createSelectorByUserOrType()  e.g. (UserID, Type, ReadOn, Created).   I imagine that this query is the most common compared to insert/delete/update.

**Review by labkey-matthewb — APPROVED** [review]
This change is better than the current behavior.  I would also consider whether a change to the notifications table is warranted.

File: core/src/org/labkey/core/notification/NotificationServiceImpl.java (hunk 1 above)
Comment (labkey-jeckels) [reply core/src/org/labkey/core/notification/NotificationServiceImpl.java]: Thanks. Updated, and I had missed that callers may pass in a Container (none currently do), so segmented the cache to handle that.

File: core/src/org/labkey/core/notification/NotificationServiceImpl.java (hunk 2 above)
Comment (labkey-jeckels) [reply core/src/org/labkey/core/notification/NotificationServiceImpl.java]: I can split off a separate issue for adding an index. I'm hesitant to add indices without a good local test scenario to make sure it gets used

**Pull request merged**